substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/JavaNetNetUtil.java

```diff
@@ -223,6 +223,114 @@ static InetAddress NET_SockaddrToInetAddress(Socket.sockaddr him, CIntPointer po
         return iaObj;
     }
 
+    // 255 JNIEXPORT jint JNICALL
+    // 256 NET_SockaddrEqualsInetAddress(JNIEnv *env, struct sockaddr *him, jobject iaObj)
+    static boolean NET_SockaddrEqualsInetAddress(Socket.sockaddr him, InetAddress iaObj) {
+        // 258 jint family = AF_INET;
+        int family = Socket.AF_INET();
+
+        int sa_family = Socket.AF_INET();
+
+        /* Restructured due to #ifdef mixing with either an else-branch or just a curlied block */
+
+        // 260 #ifdef AF_INET6
+        if (IsDefined.socket_AF_INET6()) {
+            // 261 family = getInetAddress_family(env, iaObj) == IPv4? AF_INET : AF_INET6;
+            family = getInetAddress_family(iaObj) == Target_java_net_InetAddress.IPv4 ? Socket.AF_INET() : Socket.AF_INET6();
+            // 262 JNU_CHECK_EXCEPTION_RETURN(env, JNI_FALSE);
+            // 263 if (him->sa_family == AF_INET6) {
+            sa_family = him.sa_family();
+        }
+
+        /* This only occurs if AF_INET6 above && him is IPv6 */
+        if (him.sa_family() == Socket.AF_INET6()) {
+            // 264 #ifdef WIN32
+            // 265 struct SOCKADDR_IN6 *him6 = (struct SOCKADDR_IN6 *)him;
+            // 266 #else
+            // 267 struct sockaddr_in6 *him6 = (struct sockaddr_in6 *)him;
+            NetinetIn.sockaddr_in6 him6 = (NetinetIn.sockaddr_in6) him;
+            // 268 #endif
+            // 269 jbyte *caddrNew = (jbyte *)&(him6->sin6_addr);
+            CCharPointer caddrNew = him6.sin6_addr().s6_addr();
+            // 270 if (NET_IsIPv4Mapped(caddrNew)) {
+            if (isIPv4Mapped(caddrNew)) {
+                // 271 int addrNew;
+                int addrNew;
+                // 272 int addrCur;
+                int addrCur;
+                // 273 if (family == AF_INET6) {
+                if (family == Socket.AF_INET6()) {
+                    // 274 return JNI_FALSE;
+                    return false;
+                }
+                // 276 addrNew = NET_IPv4MappedToIPv4(caddrNew);
+                addrNew = JavaNetNetUtilMD.NET_IPv4MappedToIPv4(caddrNew);
+                // 277 addrCur = getInetAddress_addr(env, iaObj);
+                addrCur = JavaNetNetUtilMD.getInetAddress_addr(iaObj);
+                // 278 JNU_CHECK_EXCEPTION_RETURN(env, JNI_FALSE);
+                // 279 if (addrNew == addrCur) {
+                if (addrNew == addrCur) {
+                    // 280 return JNI_TRUE;
+                    return true;
+                } else {
+                    // 282 return JNI_FALSE;
+                    return false;
+                }
+                // 284 } else {
+            } else {
+                // 285 jbyteArray ipaddress;
+                byte[] ipaddr;
+                // 286 jbyte caddrCur[16];
+                CCharPointer caddrCur = StackValue.get(16);
```

Comment:
Technically this should be `StackValue.get(16, CCharPointer.class)`.  Not worth a whole new pull request.